### PR TITLE
[Persistence] PrePersistence deprecation & Block lifecycle improvements (Issue #128 && Issue #105)

### LINKED ISSUE #104: [Persistence] V1 Persistence Foundation
# Objective

A basic SQL-based implementation of the Persistence module specification to enable the development of the rest of the Pocket Node.

# Origin Document
[Pocket protocol persistence specification](https://github.com/pokt-network/pocket-network-protocol/tree/main/persistence)

## Goals / Deliverables
- [x] Implementation of the persistence module
- [x] Actor schema specification
- [x] Actor query specification
- [x] Dockerized infrastructure to run a `LocalNet` with the new persistence module implementation
- [ ] Deletion / deprecation of the PrePersistence module
- [ ] Loading of some sort of state from local disc when a LocalNet node starts up
- [ ] Unit Tests & with the 1st iteration of an accompanying unit test library
- [ ] Documentation
    - [ ] Module specific README
    - [ ] Module specific CHANGELOG
    - [ ] Module specific code architecture (text and/or diagram)
    - [ ] Instructions on how to run/test/debug the module implementation
    - [ ] Global documentation / references updated

## Non-goals
- A comprehensive and complete "block store" mechanism
- Deployment of the node to a non-local environment 
- Data integrity verification and guarantees 
- Implementation/adoption of a Merkle Tree
- Implementation/adoption of a Key-Value Store

##  Testing Methodology
LocalNet and Unit Tests. See [./docs/development/README.md](docs/development/README.md) for more details.

Owners: @andrewnguyen22  @Olshansk

### LINKED ISSUE #105: [Persistence] Deprecate PrePersistence
# Objective
Deprecate and remove all PrePersistence related code.

# Origin Document
[[Persistence] V1 Persistence Foundation Integration Issue](https://github.com/pokt-network/pocket/issues/104)

# Goals
## Deliverables
- [ ] Deprecation of the PrePersistence Module and all related code, docs, resources, etc...

## Testing Methodology
LocalNet and Unit Tests. See [./docs/development/README.md](https://github.com/pokt-network/pocket/issues/docs/development/README.md) for more details.

## Non-goals
- List of goals that this issue won't address

Creator: @Olshansk 
Co-Owners: @andrewnguyen22

### LINKED ISSUE #128: [Persistence] Enable multiple persistence contexts on PostgresContext
# Objective
Enable the lifecycle of block processing and production by creating a transactional architecture around persistence contexts instead of writing directly to the database on every operation.

# Origin Document
### Background
Multiple persistence contexts enable a proper lifecycle for validators and fullnodes as they maintain 'ephemeral' states in situations like block validation/processing and block production.

# Goals / Deliverables
- [ ] Deliver completed code
- [ ] Document the architecture and intended use
- [ ] Test the transactional nature within the persistence module

## General milestone checklist
- [ ] Update all the relevant CHANGELOGs
- [ ] Update all the relevant READMEs
- [ ] Update the source code tree explanation
- [ ] Add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [ ] Create followup milestones + issues
- [ ] Document small TODO along the way

---

Creator: @andrewnguyen22

### PR DESCRIPTION
**Pull request by andrewnguyen22**
# Objective
closes #128 
- Enable the lifecycle of `block` processing and production by creating a transactional architecture around persistence contexts instead of writing directly to the database on every operation.

closes #105  and closes #104
- Deprecate and remove all `PrePersistence` related code.

# Origin Document
### Background

#128
_Multiple persistence contexts enable a proper lifecycle for validators and full nodes as they maintain 'ephemeral' states in situations like block validation/processing and block production._

Pretty early on - it was discovered that though ephemeral states are needed, there seems to be little validity in having multiple `write` contexts in parallel - rather enable write contexts sequentially (only one at a time) and enable multiple read contexts. This is currently not strictly enforced, rather needs discussion on where to delegate this responsibility (Consensus, Utility, or Persistence module).

#105 
[[Persistence] V1 Persistence Foundation Integration Issue](https://github.com/pokt-network/pocket/issues/104)

# Goals / Deliverables
- [x] Deliver completed code
- [ ] Document the architecture and intended use
- [x] Test the transactional nature within the persistence module
- [x] Deprecation of the PrePersistence Module and all related code, docs, resources, etc...

## General milestone checklist
- [x] Update all the relevant CHANGELOGs
- [x] Update all the relevant READMEs
- [ ] Update the source code tree explanation
- [ ] Add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [x] Create followup milestones + issues
- [x] Document small TODO along the way

## Follow-up Work

All follow-up work is being tracked in https://github.com/pokt-network/pocket/issues/149

---

## Checklist
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] If applicable, I have made corresponding changes to related or global README
- [ ] If applicable, I have added added new diagrams using [mermaid.js](https://mermaid-js.github.io)
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works

--- 

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>
Co-authored-by: Andrew Nguyen <andrewnguyen@Andrews-MacBook-Pro-2.local>
Co-authored-by: Andrew Nguyen <andrewnguyen@Andrews-MBP-2.lan>
Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

### CODE REVIEW

**Comment by andrewnguyen22** [comment]
A ton of conflicts here. Will work on this in parallel - but wanted to get the PR up to get some eyes on immediately

**Comment by Olshansk** [comment]
@andrewnguyen22 It's a bit hard to review the changes in isolation because of all the conflicts with master so I think you'll need to do "merge with master" (i.e. I don't suggest a rebase here) and likely spend an hour or two resolvings. I acknowledge this isn't fun, but I think it's worth it and we'll have smaller PRs as the MVP takes shape make these less painful.
--- 

I looked at the changelogs and the interface changes and my high level understanding is this: Instead of writing to the postgres DB, we read/write to a transaction and commit it at the very end.

If that's the case, at a high level, it looks good to me and will review in depth once ready.

**Comment by andrewnguyen22** [comment]
Merged with main, conflicts gone.

Added two follow up issues #148 and #149

File: persistence/genesis.go
Comment (Olshansk) [inline persistence/genesis.go:13]: In my previous implementation in populating the genesis DB I had a couple things that I don't see here.

1. Creating an account for each value
2. Using the genesis stake/value for all validators to determine the value in the pool (for consistency)

I copy-pasted a small code snippet below for reference but wanted to know whether this business logic is incorrect?

```go
        poolValues := make(map[string]*big.Int, 0)

	addValueToPool := func(poolName string, valueToAdd string) error {
		value, err := types.StringToBigInt(valueToAdd)
		if err != nil {
			return err
		}
		if present := poolValues[poolName]; present == nil {
			poolValues[poolName] = big.NewInt(0)
		}
		poolValues[poolName].Add(poolValues[poolName], value)
		return nil
	}

	for _, v := range state.Validators {
		// DISCUSS_IN_THIS_COMMIT(drewskey): If the validator (and other actors) have a balance, do we expect that
		// to be in the `accounts` type corresponding to the same address?
		if err := ctx.SetAccountAmount(v.Address, "0"); err != nil {
			return err
		}
		if err := ctx.InsertValidator(v.Address, v.PublicKey, v.Output, v.Paused, int(v.Status), v.ServiceUrl, v.StakedTokens, v.PausedHeight, v.UnstakingHeight); err != nil {
			return err
		}
		if err := addValueToPool(genesis.ValidatorStakePoolName, v.StakedTokens); err != nil {
			return err
		}
	}
```

File: persistence/genesis.go
Comment (Olshansk) [inline persistence/genesis.go:62]: I haven't taken a stab at this yet, but I do believe there's an opportunity to genercize this. Potentially a pattern like this:

```go
func (p PostgresContext)[A any] GetAllActors(height int64) (actors []Actor, err error) { 
...
acc := new(Actor)
```

Can you take a first stab/thought at it and if not, I can take a personal attempt at implementing it with a smaller code footprint?

File: persistence/module.go
Comment (Olshansk) [inline persistence/module.go:21]: NIT: please make this the first element in the struct.

File: persistence/schema/base_actor.go
Comment (Olshansk) [inline persistence/schema/base_actor.go:15]: What is the interface version?

File: persistence/schema/gov.go
Comment (Olshansk) [inline persistence/schema/gov.go:255]: I think we should revert this since it is being fixed in the other PR.

File: persistence/test/account_test.go
Comment (Olshansk) [inline persistence/test/account_test.go:19]: Why did we switch this back from `testPostgresDB` to `PostgresDB`.

My reasoning for the other was two-fold:

1. It doesn't need to be exposed outside the test module -> hence lowercase.
2. It's a database used for testing, so I made the clear & explicit.

File: persistence/test/setup_test.go
Comment (Olshansk) [inline persistence/test/setup_test.go:7]: Can you refactor this so that the package path is `github.com/pokt-network/pocket/shared/tests/` given that they are shared test utilities?

It's weird that persistence test are importing utility test helper modules.

File: persistence/test/setup_test.go
Comment (Olshansk) [inline persistence/test/setup_test.go:47]: Per my other comment, I think these should we lowercased (not exposed outside this module) and prefixed with `test` for explicitness.

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:10]: Do we not need to expose `NewRContext`?

**Review by Olshansk — CHANGES_REQUESTED** [review]
I took a first pass at this PR, but will do a more detailed one after we address some of the comments and discuss some of the open questions I left inline.

@andrewnguyen22 I also left a couple of open-ended refactoring comments but didn't have time to tackle the changes myself. If you find that it's too much of a waste of time or too out of scope, feel free just to leave a TODO or comment and I can address it in a follow-up PR.

File: persistence/db.go
Comment (andrewnguyen22) [reply persistence/db.go:29]: PostgresDB is just Tx and Blockstore, so if you'd prefer to remove that level of abstraction, it would look like this:

```
type PostgresContext struct {
	Height int64
	Tx         pgx.Tx
	Blockstore kvstore.KVStore	
}
```

Not

```
type PostgresContext struct {
	Height int64
	DB     PostgresDB
	Tx         pgx.Tx
	Blockstore kvstore.KVStore	
}
```

Comment (andrewnguyen22) [reply persistence/db.go:29]: If you'd like this done, let's push it to #149

File: persistence/schema/account.go
Comment (andrewnguyen22) [reply persistence/schema/account.go:68]: I'm not clear on the alternative - so I'm going to consider this a non-actionable for now

File: persistence/gov.go
Comment (andrewnguyen22) [reply persistence/gov.go:11]: It will go in after I think

File: persistence/shared_sql.go
Comment (andrewnguyen22) [reply persistence/shared_sql.go:15]:
```
	UnstakedStatus = iota
	UnstakingStatus 
	StakedStatus
```

This would result in a unique but non-deterministic number set so this won't work.

Fine to change the order though

File: shared/tests/utility_module/transaction_test.go
Comment (andrewnguyen22) [reply shared/tests/utility_module/transaction_test.go:29]: Because the merge and it's pretty much the same code with a slightly smaller footprint. Can push to #149

File: shared/tests/utility_module/transaction_test.go
Comment (andrewnguyen22) [reply shared/tests/utility_module/transaction_test.go:51]: These comments contain TODO's & demonstrate how the test used to work before. I want to leave it as is

File: shared/tests/utility_module/transaction_test.go
Comment (andrewnguyen22) [reply shared/tests/utility_module/transaction_test.go:35]: Love it; Pushing to #149 to not block

File: persistence/genesis.go
Comment (andrewnguyen22) [reply persistence/genesis.go:13]: The pool piece makes sense, but the account piece doesn't matter. Validators are not expected to have accounts associated with them just because they exist

Comment (andrewnguyen22) [reply persistence/genesis.go:13]: It's also expected in some genesis files for this logic to be pre-populated by the creator... That's why the pool field will likely exist in the genesis object

File: persistence/genesis.go
Comment (andrewnguyen22) [reply persistence/genesis.go:62]: I would highly encourage us to wait until we have this:
`// TODO (Andrew) genericize with actors interface once merged with #111`

This way we don't even need generics :)

Let me know if this is an acceptable placeholder

File: persistence/schema/base_actor.go
Comment (andrewnguyen22) [reply persistence/schema/base_actor.go:15]: See #111 for more of an example

Comment (andrewnguyen22) [reply persistence/schema/base_actor.go:15]:
```
type Actor interface {
	GetAddress() []byte
	GetPublicKey() []byte
	GetPaused() bool
	GetStatus() int32
	GetChains() []string
	GetGenericParam() string
	GetStakedTokens() string
	GetPausedHeight() int64
	GetUnstakingHeight() int64
	GetOutput() []byte
}
```

File: persistence/genesis.go
Comment (andrewnguyen22) [reply persistence/genesis.go:62]: #149 ^^

File: persistence/shared_sql.go
Comment (andrewnguyen22) [reply persistence/shared_sql.go]: Order changed with [d2edd36](https://github.com/pokt-network/pocket/pull/140/commits/d2edd36ddba729097db8ddbe6b82a26a3f64c4f4)

File: persistence/genesis.go
Comment (andrewnguyen22) [reply persistence/genesis.go:28]: Added with [d2edd36](https://github.com/pokt-network/pocket/pull/140/commits/d2edd36ddba729097db8ddbe6b82a26a3f64c4f4)

File: persistence/shared_sql.go
Comment (Olshansk) [reply persistence/shared_sql.go]: Note that this isn't a blocker for the PR but I want to understand why you said it's a `uniqute but non-deterministic number`.

My understanding of `iota` in Golang is that it is an enumeration that increments by 1 based on order per the screenshot and reference below.


Screen Shot 2022-08-11 at 1 19 06 PM
Reference: https://www.gopherguides.com/articles/how-to-use-iota-in-golang


I ran `make test_all` after replacing your updated code:

```go
const (
	UndefinedStakingStatus = 0
	UnstakingStatus        = 1
	StakedStatus           = 2
	UnstakedStatus         = 3
)
```

with:

```go
const (
	UndefinedStakingStatus = iota
	UnstakingStatus
	StakedStatus
	UnstakedStatus
)
```
and everything passed.

Wanted to confirm this is okay before I push or if I'm missing something?

File: persistence/db.go
Comment (Olshansk) [reply persistence/db.go:29]: Sgtm. Added this as an explicit deliverable:

Screen Shot 2022-08-11 at 1 27 18 PM

File: shared/tests/utility_module/transaction_test.go
Comment (Olshansk) [reply shared/tests/utility_module/transaction_test.go:29]: Going to handle it in this PR using VSCOde regex replace foo:

Screen Shot 2022-08-11 at 1 53 43 PM

File: shared/tests/utility_module/transaction_test.go
Comment (Olshansk) [reply shared/tests/utility_module/transaction_test.go:35]: Update the description in https://github.com/pokt-network/pocket/issues/149

File: persistence/shared_sql.go
Comment (andrewnguyen22) [reply persistence/shared_sql.go]: > a unique but non-deterministic number

Admittedly, I wasn't aware that iota always started with zero and incremented by one.
*I thought it was guaranteed unique, but not always starting at zero* which appears to be incorrect.

With that being said, let me state my case about why I don't think Iota's are a good choice for this:
- There are only 3 or 4 possible status' so it's not a huge verbosity issue
- The syntax sugar of the iota is not worth the language specific rules that come along with the change, example:
  - Code order of constants matters
  - Iotas start with zero or iota+1 starts with 1 
  - Careful with creating two iotas in the same const() block
Screen Shot 2022-08-11 at 4 42 37 PM

vs

Screen Shot 2022-08-11 at 4 47 05 PM


- **TL;DR** I still like being explicit considering how a change like this affects the app hash and other language implementations will need to match these constant values exactly.

File: shared/tests/utility_module/transaction_test.go
Comment (andrewnguyen22) [reply shared/tests/utility_module/transaction_test.go:29]: What a boss xd

File: persistence/shared_sql.go
Comment (andrewnguyen22) [reply persistence/shared_sql.go]: @Olshansk noticed you still made the change to iota. Lmk if you want to take it offline

**Comment by Olshansk** [comment]
@andrewnguyen22 I've spent most of the day going over this PR and the changes have become big enough that I've moved it into a separate branch for now.

I think there are some foundational business logic issues, but wanted to know your preference on:
1. Merging this as is and reviewing / discussing https://github.com/pokt-network/pocket/pull/165 separately
2. Waiting for me to fix https://github.com/pokt-network/pocket/pull/165 and pushing those changes to this branch

I would personally prefer (2), but am okay with (1) in order to unblock other work in parallel

**Comment by andrewnguyen22** [comment]
> @andrewnguyen22 I've spent most of the day going over this PR and the changes have become big enough that I've moved it into a separate branch for now.
> 
> I think there are some foundational business logic issues, but wanted to know your preference on:
> 
> 1. Merging this as is and reviewing / discussing [Issue #128 fixes #165](https://github.com/pokt-network/pocket/pull/165) separately
> 2. Waiting for me to fix [Issue #128 fixes #165](https://github.com/pokt-network/pocket/pull/165) and pushing those changes to this branch
> 
> I would personally prefer (2), but am okay with (1) in order to unblock other work in parallel

Let's sync offline to make sure everything goes smoothly. Not closed off to either solution, but 1) does seem attractive given you've already split the PR into two

**Review by Olshansk — COMMENTED** [review]
@andrewnguyen22 As discussed offline, I'm working on the follow-ups to this in EDIT: Moving this to https://github.com/pokt-network/pocket/pull/165.

Hoping to approve this by EOD, but just reviewing to make sure I didn't lose any context.

I'm posting some of my previous comments, only one of which I believe needs an answer.

File: persistence/shared_sql.go
Comment (Olshansk) [reply persistence/shared_sql.go]: Reversed it back.

File: persistence/account.go
Comment (andrewnguyen22) [reply persistence/account.go:24]:
```
We only return this default value if err != nil, but if err != nil, then the result should not be used because we should handle the error.
```
The error returned can be `row_not_found` (pseudo) which should be translated to the default value

Comment (andrewnguyen22) [reply persistence/account.go:24]: We can handle this higher in the stack if you prefer

**Comment by andrewnguyen22** [comment]
Continued conversation here https://github.com/pokt-network/pocket/pull/165#issuecomment-1215457696

File: persistence/module.go
Comment (Olshansk) [inline persistence/module.go:82]: I think this might be the source of the issues I'm running into in the #128-fixes/improvements. 

Can we reuse the same DB connection from `Create` here?

**Comment by Olshansk** [comment]
@andrewnguyen22 I'm getting close to finishing https://github.com/pokt-network/pocket/pull/165, so what are your thoughts on merging this in so https://github.com/pokt-network/pocket/pull/139 can start resolving merging conflicts while reviewing https://github.com/pokt-network/pocket/pull/165 in parallel?

Can approve the PR if that sounds good per out offline discussion.

**Review by Olshansk — APPROVED** [review]
:shipit:

**Pull request merged**
